pkg/storegateway/series_chunks.go

```diff
@@ -398,15 +424,56 @@ func (c *loadingSeriesChunksSetIterator) Next() (retHasNext bool) {
 	// Create a batched memory pool that can be released all at once.
 	chunksPool := pool.NewSafeSlabPool[byte](chunkBytesSlicePool, chunkBytesSlabSize)
 
-	err := c.chunkReaders.load(nextSet.series, chunksPool, c.stats)
+	err = c.chunkReaders.load(nextSet.series, chunksPool, c.stats)
 	if err != nil {
-		c.err = errors.Wrap(err, "loading chunks")
-		return false
+		return nil, false, errors.Wrap(err, "loading chunks")
 	}
 
 	nextSet.chunksReleaser = chunksPool
-	c.current = nextSet
-	return true
+
+	if c.ignoreNativeHistograms {
+		nextSet.series = filterNativeHistogramChunks(nextSet.series)
+		if len(nextSet.series) == 0 {
+			nextSet.release()
+			return nil, true, nil
+		}
+	}
+
+	return &nextSet, false, nil
+}
+
+func filterNativeHistogramChunks(seriesEntries []seriesEntry) []seriesEntry {
+	seriesWriteIdx := 0
+	for i, series := range seriesEntries {
+		chksWriteIdx := 0
+		for j, chk := range series.chks {
+			// The bucketChunkReader will skip populating native histogram chunks when ignoreNativeHistograms is true,
+			// leaving chk.Raw as nil.
+			if chk.Raw == nil {
+				continue
+			}
+
+			series.chks[chksWriteIdx] = series.chks[j]
+			if series.refs != nil {
+				// Not all callers will have series.refs set, but when it is set we expect refs and chks to have a
+				// one-to-one relationship.
+				series.refs[chksWriteIdx] = series.refs[j]
+			}
+
+			chksWriteIdx++
+		}
+		series.chks = series.chks[:chksWriteIdx]
+
+		if len(series.chks) == 0 {
+			seriesEntries[i] = seriesEntry{}
```

Comment:
Setting this to an empty element means that we'll also not be able to iterate the chunks in this slice and reset them in `seriesChunksSet.release`. I don't think this zero-ing here is correct

Reply:
The same reasoning applies to the `seriesEntries[seriesWriteIdx] = series` below - the that used to be in `seriesEntries[seriesWriteIdx]` won't be reset now. 

One way to ensure this is correct is to just swap the elements in `seriesEntries`.

Reply:
Ack, nice catch. My brain's getting a bit twisted on all of this, but I think this might've made sense when I was resetting chunks as I was filtering instead of now doing it all in `seriesChunksSet.release` and I didn't think through all the ramifications of making that change.  Anyways, I'll fix it up!

Reply:
so basically in order to reset an aggr chunk (wherever this may happen), the chunk still needs to be somehow accessible. The way it's accessed right now is via each seriesEntry element. If we zero the seriesEntry, then the chunk is "unreachable" and can't be reset.